### PR TITLE
feat(cloudweb): allow upgrading offer

### PR DESCRIPTION
**Pull request by euhmeuh**
MBP-443
- [x] Awaiting content texts (CON-1561)

### CODE REVIEW

**Comment by euhmeuh** [comment]
@cbourgois It's a peer dep of oui-timepicker

File: client/app/hosting/offer/upgrade/hosting-offer-upgrade.html [hunk 1]
```diff
@@ -1,114 +1,90 @@
...
+               data-on-dismiss="resetAction()"
+               data-loading="$ctrl.loading.availableOffers">
+        <oui-message
+          data-type="warning"
+          data-ng-if="isEmpty($ctrl.availableOffers)"
+          data-dismissable="false">
+          <span data-translate="hosting_order_upgrade_no_offer"></span>
+        </oui-message>
+
+        <oui-stepper data-ng-if="!isEmpty($ctrl.availableOffers)"
```
Comment (antleblanc) [inline client/app/hosting/offer/upgrade/hosting-offer-upgrade.html]: Prefer using `<oui-stepper>` outside of `<oui-modal>` (if possible).

File: client/app/hosting/translations/Messages_fr_FR.xml [hunk 2]
```diff
@@ -1196,6 +1198,17 @@
...
+   <translation id="hosting_order_upgrade_modal_header">Changer d'offre</translation>
+   <translation id="hosting_order_upgrade_step_header_offer">Offre</translation>
+   <translation id="hosting_order_upgrade_step_header_duration">Durée</translation>
+   <translation id="hosting_order_upgrade_step_header_start_time">Heure de la mise à jour</translation>
+   <translation id="hosting_order_upgrade_step_header_contracts">Contrats</translation>
+   <translation id="hosting_order_upgrade_start_time_label">Sélectionnez l'heure de la mise à jour</translation>
+   <translation id="hosting_order_upgrade_start_time_explain">L'heure que vous choisissez correspond à la prochaine fois que cette heure sera atteinte dans votre fuseau horaire. Si vous choisissez une heure qui est déjà passée, la mise à jour s'effectuera demain à cette heure.</translation>
+   <translation id="hosting_order_upgrade_start_time_info">Le changement d'offre Cloudweb entraine une coupure de service d'environ 15 minutes maximum après le début de l'opération. Vous pouvez donc choisir un horaire de nuit afin de limiter l'impact sur vos sites web.</translation>
```
Comment (antleblanc) [inline client/app/hosting/translations/Messages_fr_FR.xml]:
```suggestion
   <translation id="hosting_order_upgrade_start_time_info">Le changement d'offre Cloud Web entraine une coupure de service d'environ 15 minutes maximum après le début de l'opération. Vous pouvez donc choisir un horaire de nuit afin de limiter l'impact sur vos sites web.</translation>
```

Is that make sense?

Comment (euhmeuh) [reply client/app/hosting/translations/Messages_fr_FR.xml]: Content is in the process of reviewing those texts, we'll see what they come up with.

File: client/app/hosting/offer/upgrade/hosting-offer-upgrade.html (hunk 1 above)
Comment (euhmeuh) [reply client/app/hosting/offer/upgrade/hosting-offer-upgrade.html]: Yes I'm gonna kill the modal and put that in a real page.

File: client/app/css/source.less [hunk 3]
```diff
@@ -4,6 +4,7 @@
 @oui-path: "ovh-ui-kit";
 
 // Vendors
+@import (less) "flatpickr/dist/flatpickr.min.css";
```
Comment (marie-j) [inline client/app/css/source.less]: Vendors could probably be imported in `index.js` if you don't need to use variables or mixin in your sources

File: client/app/hosting/offer/upgrade/hosting-offer-upgrade.html (hunk 1 above)
Comment (euhmeuh) [reply client/app/hosting/offer/upgrade/hosting-offer-upgrade.html]: After discussing that in weekly: We're going to prod that as-is and move it to a page in a further release.

File: client/app/hosting/hosting.service.js [hunk 4]
```diff
@@ -590,13 +590,12 @@
...
+             * @param {string} startTime (optional)
              */
-      orderUpgrade(serviceName, offer, duration) {
+      orderUpgrade(serviceName, offer, duration, startTime = null) {
         return this.OvhHttp.post(`/order/hosting/web/${serviceName}/upgrade/${duration}`, {
           rootPath: 'apiv6',
-          data: {
-            offer,
-          },
+          data: Object.assign({ offer }, startTime ? { startTime } : {}),
```
Comment (FredericEspiau) [inline client/app/hosting/hosting.service.js]:
```suggestion
          data: {
              ...offer
              startTime: startTime ? startTime : {}
          },
```
I like this way better :)

File: client/app/hosting/offer/upgrade/hosting-offer-upgrade.controller.js [hunk 5]
```diff
@@ -15,114 +15,108 @@ angular.module('App').controller(
...
-        })
-        .catch((err) => {
-          this.Alerter.alertFromSWS(this.$translate.instant('hosting_order_upgrade_error'), err, this.$scope.alerts.main);
-        })
-        .finally(() => {
-          this.loading.validation = false;
-          this.$scope.resetAction();
+      const startTime = moment(`T${this.model.startTime}`).utc().format('HH:mm:ss');
+
+      return this.Hosting.orderUpgrade(
```
Comment (FredericEspiau) [inline client/app/hosting/offer/upgrade/hosting-offer-upgrade.controller.js]:
```suggestion
      return this.Hosting
          .orderUpgrade(
              ...
```
would look nicer I think

File: client/app/hosting/hosting.service.js (hunk 4 above)
Comment (euhmeuh) [reply client/app/hosting/hosting.service.js]: I think you mean:
```
data: {
  offer,
  ...(startTime ? { startTime } : {})
},
```

**Comment by ovh-cds** [comment]
CDS Report build-manager#1459.0 ✘  
* Stage 1
  * Build dependencies and inject Sentry ✔ 
* Stage 2
  * Build manager CA ✔ 
  * Build manager EU ✘  
  * Build manager US ✔ 
  * Tests ✔

File: client/app/hosting/hosting.service.js (hunk 4 above)
Comment (FredericEspiau) [reply client/app/hosting/hosting.service.js]: Let's find a compromise:
````js
data: {
  offer,
  startTime:  startTime || {},
}
````
😅

**Comment by antleblanc** [comment]
@euhmeuh  Is there any internal ticket reference related to this pull request?

Thanks!

File: client/app/hosting/offer/upgrade/hosting-offer-upgrade.html (hunk 1 above)
Comment (antleblanc) [reply client/app/hosting/offer/upgrade/hosting-offer-upgrade.html]: Could you please create a ticket which target theses improvements then?

Thanks!

**Comment by jleveugle** [comment]
Hello @euhmeuh,
Your pull request is now closed due to inactivity.
Feel free to re-open it if you have any update.
Anna Bot for OVH UX <3

File: client/app/hosting/offer/upgrade/hosting-offer-upgrade.controller.js [hunk 6]
```diff
@@ -15,113 +16,116 @@ angular.module('App').controller(
...
+      const time = moment(`2000-01-01T${startTime}`);
+      const now = moment(moment().format('2000-01-01THH:mm'));
+      const isToday = time.diff(now) > 0;
 
-    /* Step 4 */
-    backToContracts() {
-      if (!this.model.duration.contracts || !this.model.duration.contracts.length) {
-        this.$rootScope.$broadcast('wizard-goToStep', 3);
-      }
+      return isToday
```
Comment (FredericEspiau) [inline client/app/hosting/offer/upgrade/hosting-offer-upgrade.controller.js]:
```suggestion
      return this.$translate.instant(
        isToday 
          ? 'hosting_order_upgrade_start_time_summary_today'
          : 'hosting_order_upgrade_start_time_summary_tomorrow', 
        { time: time.format('LT (UTCZ)') },
      )
```
would be nicer imo :)

File: client/app/hosting/offer/upgrade/hosting-offer-upgrade.html [hunk 7]
```diff
@@ -1,114 +1,108 @@
...
-                        <span class="oui-radio__label">
-                            <span data-ng-bind="duration.duration | wucDuration: 'longDate'"></span>&nbsp;:
-                            <span data-ng-bind-html="duration.prices | price: ctrl.ovhSubsidiary"></span>
-                        </span>
-                    </label>
+<div class="row tab-content"
+     data-ng-controller="HostingUpgradeOfferCtrl as $ctrl">
+    <div class="col-md-8">
+        <oui-back-button data-sref="^">
+            {{:: 'hosting_order_upgrade_modal_header' | translate }}
```
Comment (FredericEspiau) [inline client/app/hosting/offer/upgrade/hosting-offer-upgrade.html]:
```suggestion
            <span data-translate="hosting_order_upgrade_modal_header"></span>
```

File: client/app/hosting/offer/upgrade/hosting-offer-upgrade.controller.js [hunk 8]
```diff
@@ -86,10 +86,15 @@ angular.module('App').controller(
         });
     }
 
+    static convertTimeToMoment(stringTime) {
+      const now = moment();
+      const [hours, minutes, seconds] = stringTime.split(':').map(s => parseInt(s, 10));
+      return now.hours(hours).minutes(minutes).seconds(seconds);
+    }
```
Comment (marie-j) [inline client/app/hosting/offer/upgrade/hosting-offer-upgrade.controller.js]: `moment(stringTime, 'HH::mm:ss')` won't do ?

Comment (euhmeuh) [reply client/app/hosting/offer/upgrade/hosting-offer-upgrade.controller.js]: Sigh... Thanks...

File: client/app/hosting/module/MODULE.html [hunk 9]
```diff
@@ -20,7 +20,7 @@
                     <p data-translate="hosting_change_offer_for_service_start10m"></p>
                     <button class="btn btn-info" type="button"
                             data-translate="hosting_dashboard_service_change_offer"
-                            data-ng-click="setAction('offer/upgrade/hosting-offer-upgrade')">
+                            data-ng-click="ctrlHostingTabModules.$state.go('app.hosting.upgrade', { 'productId': hosting.serviceName })">
```
Comment (antleblanc) [inline client/app/hosting/module/MODULE.html]: Is `ui-sref` directive couldn't suite your needs?

File: client/app/hosting/database/add/hosting-database-add.html [hunk 10]
```diff
@@ -62,15 +62,15 @@
                     <li data-ng-if="isPerso()">
                         <button class="btn btn-link text-left text-wrap p-0" type="button"
                                 data-translate="hosting_tab_DATABASES_configuration_create_upgrade_perso"
-                                data-ng-click="setAction('offer/upgrade/hosting-offer-upgrade', hosting)">
+                                data-ng-click="goToHostingUpgrade()">
```
[the remark this answers is not in the document]
Comment (euhmeuh) [reply client/app/hosting/database/add/hosting-database-add.html]: Didn't knew you could pass objects to that syntax. Thanks!

**Pull request merged**